components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt

```diff
@@ -234,18 +259,27 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                                     )
                                     val client = connector.connect()
                                     client.clientClosed.advise(connectionLifetime) {
+                                        thisLogger().d("clientClosed, ${client.prettyPrint()}")
                                         application.invokeLater {
+                                            connectionLifetime.onTermination {
+                                                thisLogger().d("clientTermination, clientClosed didChange fired, ${client.prettyPrint()}")
+                                                dispatcher.multicaster.didChange()
+                                            }
                                             connectionLifetime.terminate()
                                         }
                                     }
```

Comment:
One thing I've noticed, is that we could benefit from letting the `connectionLifetime.onTermination` set even before the client gets closed. Because there is another place where we call `connectionLifetime.terminate()`, and in the future we can have even more cases.
Something like the following, which can be done in a follow-up PR if needed:
```suggestion
                                    connectionLifetime.onTermination {
                                        thisLogger().d("clientLifetimeTermination, didChange fired, ${client.prettyPrint()}")
                                        dispatcher.multicaster.didChange()
                                    }
                                    client.clientClosed.advise(connectionLifetime) {
                                        thisLogger().d("clientClosed, ${client.prettyPrint()}")
                                        application.invokeLater {
                                            connectionLifetime.terminate()
                                        }
                                    }
```